File: apps/server-admin-view/src/lib/api/wol.ts
import type { components as ApiContractComponents } from "@fn-knock/api-contract";

import { apiClient } from "./client";

type WolSchemas = ApiContractComponents["schemas"];

export type WOLRelay = WolSchemas["WolRelayData"];
export type WOLRelaySummary = WolSchemas["WolRelaySummaryData"];
export type WOLTarget = WolSchemas["WolTargetData"];
export type WOLIntegrationRuntime =
  WolSchemas["WolIntegrationRuntimeData"];
export type WOLIntegrationRuntimeState = WOLIntegrationRuntime["state"];
export type WOLTargetIntegrations =
  WolSchemas["WolTargetIntegrationsData"];
type WOLBlinkerIntegrationInput =
  Omit<WolSchemas["WolBlinkerIntegrationInputData"], "deviceKey"> &
    Required<
      Pick<
        WolSchemas["WolBlinkerIntegrationInputData"],
        "bindComponent" | "skipTlsVerify"
      >
    > & { deviceKey?: string };
type WOLBemfaIntegrationInput =
  Omit<WolSchemas["WolBemfaIntegrationInputData"], "privateKey"> &
    Required<
      Pick<
        WolSchemas["WolBemfaIntegrationInputData"],
        "topic" | "skipTlsVerify"
      >
    > & { privateKey?: string };
export type WOLTargetIntegrationInput = {
  blinker: WOLBlinkerIntegrationInput;
  bemfa: WOLBemfaIntegrationInput;
};
export type WOLTargetStatus = WolSchemas["WolTargetStatusData"];
export type WOLRelayInput = WolSchemas["WolRelayInputData"] &
  Required<Pick<WolSchemas["WolRelayInputData"], "port" | "enabled">>;
export type WOLTargetInput = Omit<
  WolSchemas["WolTargetInputData"],
  "integrations"
> &
  Required<
    Pick<
      WolSchemas["WolTargetInputData"],
      "relayId" | "broadcastAddress" | "ipAddress" | "enabled"
    >
  > & { integrations?: WOLTargetIntegrationInput };
export type WOLBootstrap = WolSchemas["WolBootstrapData"];
export type WOLRelayCredentialResult =
  WolSchemas["WolRelayCredentialData"];
export type WOLDispatchResult = WolSchemas["WolDispatchData"];
export type WOLLocalRelayConfig = WolSchemas["WolLocalRelayConfigData"];
export type WOLLocalRelayRuntime = WolSchemas["WolLocalRelayRuntimeData"];
export type WOLLocalRelay = WolSchemas["WolLocalRelayData"];
export type WOLLocalRelayInput = WolSchemas["WolLocalRelayInputData"] &
  Required<Pick<WolSchemas["WolLocalRelayInputData"], "allowedSources">>;
export type WOLLocalNetwork = WolSchemas["WolLocalNetworkData"];
export type WOLDiscoveredDevice = WolSchemas["WolDiscoveredDeviceData"];
export type WOLDiscoveryResult = WolSchemas["WolDiscoveryResultData"];
export type WOLDiscoveryProgress = WolSchemas["WolDiscoveryProgressData"];
export type WOLDiscoveryJobStatus = WolSchemas["WolDiscoveryJobData"];
export type WOLDiscoveryJobState = WOLDiscoveryJobStatus["state"];

export type WOLDiscoveryPollEvent =
  | { type: "meta"; data: WOLDiscoveryJobStatus }
  | { type: "progress"; data: WOLDiscoveryProgress }
  | { type: "device"; data: WOLDiscoveredDevice }
  | { type: "done"; data: WOLDiscoveryResult }
  | { type: "cancelled" };

export type WOLDiscoveryPollOptions = {
  signal?: AbortSignal;
  intervalMs?: number;
  onEvent?: (event: WOLDiscoveryPollEvent) => void;
};

const createDiscoveryAbortError = () => {
  const error = new Error("Discovery cancelled");
  error.name = "AbortError";
  return error;
};

const throwIfDiscoveryAborted = (signal?: AbortSignal) => {
  if (signal?.aborted) throw createDiscoveryAbortError();
};

const waitForDiscoveryPoll = (ms: number, signal?: AbortSignal) =>
  new Promise<void>((resolve, reject) => {
    if (signal?.aborted) {
      reject(createDiscoveryAbortError());
      return;
    }
    const timer = globalThis.setTimeout(() => {
      signal?.removeEventListener("abort", onAbort);
      resolve();
    }, ms);
    const onAbort = () => {
      globalThis.clearTimeout(timer);
      signal?.removeEventListener("abort", onAbort);
      reject(createDiscoveryAbortError());
    };
    signal?.addEventListener("abort", onAbort, { once: true });
  });

type WOLRelayList = WolSchemas["WolRelayListData"];
type WOLTargetList = WolSchemas["WolTargetListData"];
type WOLLocalRelayPairBody = WolSchemas["WolLocalRelayPairBodyData"];
type WOLDiscoveryBody = WolSchemas["WolDiscoveryBodyData"];

export const WOLAPI = {
  async getLocalRelay(): Promise<WOLLocalRelay> {
    const response = await apiClient.get("/wol/local-relay");
    return response.data.data;
  },
  async updateLocalRelay(payload: WOLLocalRelayInput): Promise<WOLLocalRelay> {
    const response = await apiClient.put("/wol/local-relay", payload);
    return response.data.data;
  },
  async pairLocalRelay(pairingCode: string): Promise<WOLLocalRelay> {
    const payload: WOLLocalRelayPairBody = {
      pairingCode,
    };
    const response = await apiClient.post("/wol/local-relay/pair", payload);
    return response.data.data;
  },
  async listRelays(): Promise<WOLRelayList> {
    const response = await apiClient.get("/wol/relays");
    return response.data.data;
  },
  async createRelay(payload: WOLRelayInput): Promise<WOLRelayCredentialResult> {
    const response = await apiClient.post("/wol/relays", payload);
    return response.data.data;
  },
  async updateRelay(id: string, payload: WOLRelayInput): Promise<WOLRelay> {
    const response = await apiClient.put(
      `/wol/relays/${encodeURIComponent(id)}`,
      payload,
    );
    return response.data.data;
  },
  async deleteRelay(id: string): Promise<void> {
    await apiClient.delete(`/wol/relays/${encodeURIComponent(id)}`);
  },
  async rotateRelayPsk(id: string): Promise<WOLRelayCredentialResult> {
    const response = await apiClient.post(
      `/wol/relays/${encodeURIComponent(id)}/rotate-psk`,
    );
    return response.data.data;
  },
  async probeRelay(id: string): Promise<WOLDispatchResult> {
    const response = await apiClient.post(
      `/wol/relays/${encodeURIComponent(id)}/probe`,
    );
    return response.data.data;
  },
  async listTargets(): Promise<WOLTargetList> {
    const response = await apiClient.get("/wol/targets");
    return response.data.data;
  },
  async getTarget(id: string, signal?: AbortSignal): Promise<WOLTarget> {
    const response = await apiClient.get(
      `/wol/targets/${encodeURIComponent(id)}`,
      { signal },
    );
    return response.data.data;
  },
  async startDiscoveryJob(
    targetCidrs: string[],
    signal?: AbortSignal,
  ): Promise<WOLDiscoveryJobStatus> {
    const payload: WOLDiscoveryBody = { targetCidrs };
    const response = await apiClient.post(
      "/wol/discover/jobs",
      payload,
      { signal },
    );
    return response.data.data;
  },
  async getDiscoveryJob(
    jobId: string,
    cursor: number,
    signal?: AbortSignal,
  ): Promise<WOLDiscoveryJobStatus> {
    const response = await apiClient.get(
      `/wol/discover/jobs/${encodeURIComponent(jobId)}`,
      { params: { cursor }, signal },
    );
    return response.data.data;
  },
  async cancelDiscoveryJob(jobId: string): Promise<WOLDiscoveryJobStatus> {
    const response = await apiClient.delete(
      `/wol/discover/jobs/${encodeURIComponent(jobId)}`,
    );
    return response.data.data;
  },
  async discoverLocalDevices(
    targetCidrs: string[],
    options: WOLDiscoveryPollOptions = {},
  ): Promise<WOLDiscoveryResult> {
    let jobId = "";
    let cursor = 0;
    let cancelRequested = false;
    const requestCancel = () => {
      if (!jobId || cancelRequested) return;
      cancelRequested = true;
      void this.cancelDiscoveryJob(jobId).catch(() => undefined);
    };

    try {
      throwIfDiscoveryAborted(options.signal);
      const started = await this.startDiscoveryJob(targetCidrs, options.signal);
      jobId = started.jobId;
      options.onEvent?.({ type: "meta", data: started });

      while (true) {
        throwIfDiscoveryAborted(options.signal);
        const status = await this.getDiscoveryJob(
          jobId,
          cursor,
          options.signal,
        );
        options.onEvent?.({ type: "progress", data: status.progress });
        for (const device of status.devices) {
          options.onEvent?.({ type: "device", data: device });
        }
        cursor = status.nextCursor;

        if (status.state === "completed") {
          if (!status.result)
            throw new Error("Discovery completed without a result");
          options.onEvent?.({ type: "done", data: status.result });
          return status.result;
        }
        if (status.state === "cancelled") {
          options.onEvent?.({ type: "cancelled" });
          throw createDiscoveryAbortError();
        }
        if (status.state === "failed") {
          throw new Error(status.error || "LAN discovery failed");
        }
        await waitForDiscoveryPoll(options.intervalMs ?? 350, options.signal);
      }
    } catch (error) {
      if (options.signal?.aborted) {
        requestCancel();
        throw createDiscoveryAbortError();
      }
      throw error;
    }
  },
  async createTarget(payload: WOLTargetInput): Promise<WOLTarget> {
    const response = await apiClient.post("/wol/targets", payload);
    return response.data.data;
  },
  async updateTarget(id: string, payload: WOLTargetInput): Promise<WOLTarget> {
    const response = await apiClient.put(
      `/wol/targets/${encodeURIComponent(id)}`,
      payload,
    );
    return response.data.data;
  },
  async deleteTarget(id: string): Promise<void> {
    await apiClient.delete(`/wol/targets/${encodeURIComponent(id)}`);
  },
  async wakeTarget(id: string): Promise<WOLDispatchResult> {
    const response = await apiClient.post(
      `/wol/targets/${encodeURIComponent(id)}/wake`,
    );
    return response.data.data;
  },
};

File: apps/server-admin-rs/src/infra/openapi_docs/domain_contracts/wol.rs
use serde::Serialize;
use utoipa::ToSchema;

#[derive(Serialize, ToSchema)]
pub(super) struct WolFeatureConfigData {
    enabled: bool,
}

#[derive(Serialize, ToSchema)]
pub(super) struct WolFeatureConfigUpdateData {
    #[schema(nullable = false)]
    enabled: Option<bool>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolLocalRelayConfigData {
    enabled: bool,
    relay_id: String,
    key_version: u32,
    listen_address: String,
    port: u16,
    broadcast_destinations: Vec<String>,
    allowed_sources: Vec<String>,
    psk_configured: bool,
    updated_at: String,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolLocalRelayRuntimeData {
    enabled: bool,
    active: bool,
    #[schema(required = true)]
    listen_address: Option<String>,
    #[schema(required = true)]
    last_error: Option<String>,
    #[schema(required = true)]
    updated_at: Option<String>,
}

#[derive(Serialize, ToSchema)]
pub(super) struct WolLocalRelayData {
    config: WolLocalRelayConfigData,
    runtime: WolLocalRelayRuntimeData,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolLocalRelayInputData {
    enabled: bool,
    relay_id: String,
    key_version: u32,
    listen_address: String,
    port: u16,
    broadcast_destinations: Vec<String>,
    #[schema(nullable = false)]
    allowed_sources: Option<Vec<String>>,
    psk: Option<String>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolLocalRelayPairBodyData {
    pairing_code: String,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolRelayInputData {
    name: String,
    address: String,
    #[schema(nullable = false)]
    port: Option<u16>,
    #[schema(nullable = false)]
    enabled: Option<bool>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolRelayData {
    id: String,
    name: String,
    address: String,
    port: u16,
    enabled: bool,
    key_version: u32,
    psk_configured: bool,
    created_at: String,
    updated_at: String,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolRelaySummaryData {
    id: String,
    name: String,
    address: String,
    port: u16,
    enabled: bool,
    psk_configured: bool,
}

#[derive(Serialize, ToSchema)]
pub(super) struct WolRelayListData {
    total: usize,
    items: Vec<WolRelayData>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolBootstrapData {
    pairing_code: String,
}

#[derive(Serialize, ToSchema)]
pub(super) struct WolRelayCredentialData {
    relay: WolRelayData,
    bootstrap: WolBootstrapData,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolDispatchData {
    request_id: String,
    #[schema(required = true)]
    relay_id: Option<String>,
    delivery_mode: String,
    target_id: Option<String>,
    status: String,
    attempts: u8,
    latency_ms: u64,
    acknowledged_at: String,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolTargetStatusData {
    state: String,
    #[schema(required = true)]
    checked_at: Option<String>,
    #[schema(required = true)]
    last_online_at: Option<String>,
    #[schema(required = true)]
    observed_ip: Option<String>,
    #[schema(required = true)]
    last_error: Option<String>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolIntegrationRuntimeData {
    state: String,
    #[schema(required = true)]
    last_connected_at: Option<String>,
    #[schema(required = true)]
    last_message_at: Option<String>,
    #[schema(required = true)]
    last_error: Option<String>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolBlinkerIntegrationData {
    enabled: bool,
    bind_component: bool,
    skip_tls_verify: bool,
    credential_configured: bool,
    runtime: WolIntegrationRuntimeData,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolBemfaIntegrationData {
    enabled: bool,
    topic: String,
    skip_tls_verify: bool,
    credential_configured: bool,
    runtime: WolIntegrationRuntimeData,
}

#[derive(Serialize, ToSchema)]
pub(super) struct WolTargetIntegrationsData {
    blinker: WolBlinkerIntegrationData,
    bemfa: WolBemfaIntegrationData,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolBlinkerIntegrationInputData {
    enabled: bool,
    device_key: Option<String>,
    #[schema(nullable = false)]
    bind_component: Option<bool>,
    #[schema(nullable = false)]
    skip_tls_verify: Option<bool>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolBemfaIntegrationInputData {
    enabled: bool,
    private_key: Option<String>,
    #[schema(nullable = false)]
    topic: Option<String>,
    #[schema(nullable = false)]
    skip_tls_verify: Option<bool>,
}

#[derive(Serialize, ToSchema)]
pub(super) struct WolTargetIntegrationsInputData {
    blinker: Option<WolBlinkerIntegrationInputData>,
    bemfa: Option<WolBemfaIntegrationInputData>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolTargetSshInputData {
    enabled: bool,
    host: String,
    #[schema(nullable = false)]
    port: Option<u16>,
    username: String,
    platform: String,
    auth_method: String,
    host_key_algorithm: String,
    host_key_fingerprint: String,
    password: Option<String>,
    private_key: Option<String>,
    private_key_passphrase: Option<String>,
    #[schema(nullable = false)]
    clear_credential: Option<bool>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolTargetSshData {
    enabled: bool,
    host: String,
    port: u16,
    username: String,
    platform: String,
    auth_method: String,
    host_key_algorithm: String,
    host_key_fingerprint: String,
    credential_configured: bool,
    passphrase_configured: bool,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolSshConnectionTestData {
    authenticated: bool,
    privilege_ready: bool,
    latency_ms: u64,
    host_key_algorithm: String,
    host_key_fingerprint: String,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolShutdownData {
    target_id: String,
    status: String,
    platform: String,
    latency_ms: u64,
    requested_at: String,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolTargetInputData {
    name: String,
    mac: String,
    relay_id: Option<String>,
    broadcast_address: Option<String>,
    ip_address: Option<String>,
    #[schema(nullable = false)]
    enabled: Option<bool>,
    integrations: Option<WolTargetIntegrationsInputData>,
    ssh: Option<WolTargetSshInputData>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolTargetData {
    id: String,
    name: String,
    mac: String,
    #[schema(required = true)]
    relay_id: Option<String>,
    #[schema(required = true)]
    broadcast_address: Option<String>,
    #[schema(required = true)]
    ip_address: Option<String>,
    delivery_mode: String,
    enabled: bool,
    created_at: String,
    updated_at: String,
    #[schema(required = true)]
    relay: Option<WolRelaySummaryData>,
    status: WolTargetStatusData,
    integrations: WolTargetIntegrationsData,
    ssh: WolTargetSshData,
}

#[derive(Serialize, ToSchema)]
pub(super) struct WolTargetListData {
    total: usize,
    items: Vec<WolTargetData>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolDiscoveryBodyData {
    #[schema(nullable = false)]
    target_cidrs: Option<Vec<String>>,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolLocalNetworkData {
    interface_name: String,
    address: String,
    cidr: String,
    scan_cidr: String,
    broadcast_address: String,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolDiscoveredDeviceData {
    ip: String,
    mac: String,
    interface_name: String,
    broadcast_address: String,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolDiscoveryProgressData {
    scanned_hosts: usize,
    total_hosts: usize,
    found_devices: usize,
    current_host: String,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolDiscoveryResultData {
    devices: Vec<WolDiscoveredDeviceData>,
    networks: Vec<WolLocalNetworkData>,
    duration_ms: u64,
    method: String,
}

#[derive(Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub(super) struct WolDiscoveryJobData {
    job_id: String,
    state: String,
    created_at: i64,
    updated_at: i64,
    networks: Vec<WolLocalNetworkData>,
    progress: WolDiscoveryProgressData,
    devices: Vec<WolDiscoveredDeviceData>,
    next_cursor: usize,
    #[schema(required = true)]
    result: Option<WolDiscoveryResultData>,
    #[schema(required = true)]
    error: Option<String>,
}
